support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -70,53 +60,48 @@ object GeneratedGiftCode {
 
 }
 
-sealed trait GiftPurchase {
-  def asDigitalSubscriptionGiftPurchase: Option[DigitalSubscriptionGiftPurchase] =
-    this match { case a: DigitalSubscriptionGiftPurchase => Some(a); case _ => None }
-  def giftRecipient: GiftRecipient
-}
-object GiftPurchase {
-
-  case class WeeklyGiftPurchase(giftRecipient: WeeklyGiftRecipient) extends GiftPurchase
-  case class DigitalSubscriptionGiftPurchase(
-    giftRecipient: DigitalSubscriptionGiftRecipient,
-    giftCode: GeneratedGiftCode,
-    lastRedemptionDate: LocalDate,
-  ) extends GiftPurchase
-
-  val circeDiscriminator = new CirceDiscriminator("giftPurchaseType")
-  private val discriminatorWeekly = "Weekly"
-  private val discriminatorDigitalSubscription = "DigitalSubscription"
-
-  implicit val e1 = circeDiscriminator.add(deriveEncoder[WeeklyGiftPurchase], discriminatorWeekly)
-  implicit val d1 = deriveDecoder[WeeklyGiftPurchase]
-  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubscriptionGiftPurchase], discriminatorDigitalSubscription)
-  implicit val d2 = deriveDecoder[DigitalSubscriptionGiftPurchase]
-
-  implicit val encoder: Encoder[GiftPurchase] = Encoder.instance {
-    case c: WeeklyGiftPurchase => c.asJson
-    case c: DigitalSubscriptionGiftPurchase => c.asJson
+class DiscriminatedType[TOPLEVEL](discriminatorFieldName: String) {
```

Comment:
Move this to it's own file maybe? I'd also put it in the `com.gu.support.encoding` package

Reply:
good spot 👍